### PR TITLE
Install into bare repos.

### PR DESCRIPTION
**Pull request by gabyx**
Some initial fixes to allow bare repo installs.
Tests should be added...
@rycus86 Could you do this?

### CODE REVIEW

**Comment by gabyx** [comment]
Does travis tests not run anymore?

**Comment by rycus86** [comment]
There's been some issues with Travis yesterday too, it might be a bit borked.

**Comment by rycus86** [comment]
Oh, here it is: https://travis-ci.org/rycus86/githooks/builds/607703891

**Comment by gabyx** [comment]
I added two tests. GIT_DIR is again leaking (git version 2.22), I am curious if we should safe guard this.
Thats why I needed to do `PULL_OUTPUT=$(git --work-tree="$SHARED_ROOT" --git-dir="$SHARED_ROOT/.git" pull 2>&1)`
This sucks... I did not find which command seems to leak, sometime its arbitrarily...

**Comment by gabyx** [comment]
Tests almost pass. However the whitespace test fails with the test 110. Somthing with whitespaces in shared repo url (its a local one)... but `set_shared_root` looks good... and IFS also...?

**Comment by rycus86** [comment]
> Tests almost pass. However the whitespace test fails with the test 110. Somthing with whitespaces in shared repo url (its a local one)... but `set_shared_root` looks good... and IFS also...?

It's odd. Try quoting the `$(..)` expression in `execute_local_shared_hooks` and `execute_global_shared_hooks`, see if that helps? I'd expect other tests to break too though if that would be the case.

File: base-template.sh [hunk 1]
```diff
@@ -471,7 +473,7 @@ update_shared_hooks_if_appropriate() {
 
             if [ -d "$SHARED_ROOT/.git" ]; then
                 echo "* Updating shared hooks from: $SHARED_REPO"
-                PULL_OUTPUT=$(cd "$SHARED_ROOT" && git pull 2>&1)
+                PULL_OUTPUT=$(git --work-tree="$SHARED_ROOT" --git-dir="$SHARED_ROOT/.git" pull 2>&1)
```
Comment (rycus86) [inline base-template.sh]: Perhaps `PULL_OUTPUT=$(cd "$SHARED_ROOT" && git -c core.hooksPath=/dev/null pull 2>&1)` would work too?

Comment (rycus86) [inline base-template.sh]:
```suggestion
                PULL_OUTPUT=$(cd "$SHARED_ROOT" && git -c core.hooksPath=/dev/null pull 2>&1)
```

File: cli.sh [hunk 2]
```diff
@@ -1192,7 +1189,7 @@ update_shared_hooks_in() {
 
         if [ -d "$SHARED_ROOT/.git" ]; then
             echo "* Updating shared hooks from: $SHARED_REPO"
-            PULL_OUTPUT=$(cd "$SHARED_ROOT" && git pull 2>&1)
+            PULL_OUTPUT=$(git --work-tree="$SHARED_ROOT" --git-dir="$SHARED_ROOT/.git" pull 2>&1)
```
Comment (rycus86) [inline cli.sh]: Same:
```suggestion
            PULL_OUTPUT=$(cd "$SHARED_ROOT" && git -c core.hooksPath=/dev/null pull 2>&1)
```

File: install.sh [hunk 3]
```diff
@@ -3525,6 +3525,17 @@ disable_tty_input() {
...
 
+############################################################
+# Checks whether the given directory
+#   is a Git repository (bare included) or not.
+#
+# Returns:
+#   1 if failed, 0 otherwise
+############################################################
+is_git_repo() {
+    (cd "$1" && git rev-parse --is-bare-repository >/dev/null 2>&1) || return 1
```
Comment (rycus86) [inline install.sh]: I don't understand why this didn't break tests. :thinking: 
I just run `git rev-parse --is-bare-repository` on a local repository and it returns `false`, but it is a Git repo.

File: install.sh [hunk 4]
```diff
@@ -3947,7 +3958,11 @@ install_into_existing_repositories() {
 
     git config --global githooks.previous.searchdir "$RAW_START_DIR"
 
-    LOCAL_REPOSITORY_LIST=$(find "$START_DIR" -type d -name .git 2>/dev/null)
+    LOCAL_REPOSITORY_LIST=$(
+        find "$START_DIR" \( -type d -and -name .git \) -or \
+            \( -type f -and -name HEAD -and -not -path "*/.git/*" \) 2>/dev/null
```
Comment (rycus86) [inline install.sh]: `-type f -and -name HEAD -and -not -path "*/.git/*"` that's a bit of a weak test. Can we do better to find likely repositories but discard false positives?

File: install.sh [hunk 5]
```diff
@@ -3947,7 +3958,11 @@ install_into_existing_repositories() {
...
+    )
+    echo "$START_DIR, REPOS: $LOCAL_REPOSITORY_LIST"
```
Comment (rycus86) [inline install.sh]: Looks like debug statement left in?

File: tests/step-109.sh [hunk 6]
```diff
@@ -0,0 +1,34 @@
+#!/bin/sh
+# Test:
+#   Set up bare repos, run the install and verify the hooks get installed
```
Comment (rycus86) [inline tests/step-109.sh]: Nice one!

File: tests/step-110.sh [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    echo "Neither curl nor wget is available"
+    exit 249
+fi
+
+if ! sh /var/lib/githooks/install.sh; then
+    echo "! Failed to execute the install script"
+    exit 1
+fi
+
+mkdir -p /tmp/test110/p001 &&
```
Comment (rycus86) [inline tests/step-110.sh]: Can you rename the `p001`, `p002` and `p003` directories to something more meaningful?
This test ended up somewhat long and it's quite hard to follow what is what.

File: tests/test-windows.sh [hunk 8]
```diff
@@ -33,10 +33,12 @@ sed -i -E 's|HOOK_NAME=.*|HOOK_NAME=\${HOOK_NAME:-\$(basename "\$0")}|' "$ROOT_D
 
 if [ -n "${EXTRA_INSTALL_ARGS}" ]; then
     sed -i "s|sh \"$ROOT_DIR\"/install.sh|sh \"$ROOT_DIR\"/install.sh \${EXTRA_INSTALL_ARGS}|g" /var/lib/tests/step-* || exit 6
+    # shellcheck disable=SC1117
```
Comment (rycus86) [inline tests/test-windows.sh]: Why is this suddenly needed?

File: uninstall.sh [hunk 9]
```diff
@@ -137,10 +137,7 @@ uninstall_from_existing_repositories() {
 #   1 otherwise
 #####################################################
 is_running_in_git_repo_root() {
-    if ! git status >/dev/null 2>&1; then
-        return 1
-    fi
-
+    git rev-parse --is-bare-repository >/dev/null 2>&1 || return 1
```
Comment (rycus86) [inline uninstall.sh]: Same as above, this gives me `false` for a regular git repo.

File: tests/step-110.sh [hunk 10]
```diff
@@ -0,0 +1,80 @@
+#!/bin/sh
+# Test:
+#   Trigger hooks on a bare repo with a push from a local repo.
+
+if ! curl --version && ! wget --version; then
```
Comment (rycus86) [inline tests/step-110.sh]: Why does this test need `curl` or `wget`?

File: base-template.sh (hunk 1 above)
Comment (gabyx) [reply base-template.sh]: Sure, but my changes are not due to the hooks beeing executed-> its because it needs `unset GIT_DIR` or the solution currently done.

`-c core.hooksPath=/dev/null` is certainly good

File: cli.sh (hunk 2 above)
Comment (gabyx) [reply cli.sh]: here GIT_DIR is defined.
When you run `git hooks shared update` suddenly GIT_DIR is defined (perhaps from before `git hooks` invocations...

File: install.sh (hunk 3 above)
Comment (gabyx) [reply install.sh]: Really, which git version, hm.. probably not such a good idea.

File: install.sh (hunk 4 above)
Comment (gabyx) [reply install.sh]: I could not come up with a better thing. We would need a file in the main folder of the repo:
There is a `continue` in the loop below...
```
HEAD
hooks
objects
refs
```

File: install.sh (hunk 5 above)
Comment (gabyx) [reply install.sh]: Thx!

File: tests/step-110.sh (hunk 7 above)
Comment (gabyx) [reply tests/step-110.sh]: Clear :-)

File: uninstall.sh (hunk 9 above)
Comment (gabyx) [reply uninstall.sh]: Yeah, but the point is, you only want to check if its a git repo :)
This function works because if its not a git repo (bare or not)
`git rev-parse --is-bare-repository` returns not `1`

File: install.sh (hunk 3 above)
Comment (gabyx) [reply install.sh]: Yeah, but the point is, you only want to check if its a git repo :)
This function works because if its not a git repo (bare or not)
`git rev-parse --is-bare-repository` returns not `1`

Comment (gabyx) [reply install.sh]: This should be faster, then `git status` which might really slow down things. and works for both bare and non-bare.

**Comment by gabyx** [comment]
Some Problems in `git hooks shared add ...` "the space is gone"...

image

**Comment by gabyx** [comment]
The following: `git hooks shared add "/tmp/test110/hooks" || exit 1`
results in 
image
Strange everything is quoted through the chain??

**Comment by gabyx** [comment]
Whaat? Why does this split here
git hooks shared add "/tmp/test 110/hooks" || exit 1
```
choose_command() {
    echo "DEBUG: -$1- , -$2-, -$3-, -$4- "
```
image

Is this really a git alias problem???

**Comment by gabyx** [comment]
Ahh fixed, damm -> The sed replacement is the issue. 
`git hooks shared add /tmp/test 110/hooks || exit 1` works

File: tests/test-windows.sh (hunk 8 above)
Comment (gabyx) [reply tests/test-windows.sh]: shellcheck on my machine, pre-commit hook complained

File: tests/step-110.sh (hunk 10 above)
Comment (gabyx) [reply tests/step-110.sh]: Oh I dont know... it does a update I think, can I remove this, need to look what to do for the install...

**Comment by gabyx** [comment]
> Added a bunch of comments, but the main gist is that I don't understand why the new Git root detection works (if I run the command on my machine, it returns `false` as expected - but then how would Githooks know it's a Git repo directory?)

> See the comments above.

**Comment by rycus86** [comment]
Thanks! I'll try to have a proper look this weekend.

File: tests/step-110.sh (hunk 10 above)
Comment (rycus86) [reply tests/step-110.sh]: `curl` or `wget` should only be needed if the test explicitly tests downloading something, otherwise it shouldn't be skipped just because these are missing.

File: uninstall.sh (hunk 9 above)
Comment (rycus86) [reply uninstall.sh]: I have some issues with naming here.
It says `is_running_in_git_repo_root()` but AFAICT would work anywhere inside a git repo, not just the root.

Comment (rycus86) [reply uninstall.sh]: Also, the `git rev-parse --is-bare-repository` is still a bit interesting for me.
Can you please try if a simple `git rev-parse >/dev/null 2>&1 || return 1` would do what we want here and work for both bare repos and work trees?

File: tests/test-windows.sh (hunk 8 above)
Comment (rycus86) [reply tests/test-windows.sh]: Can you please check what version of shellcheck you have? Thanks!

**Comment by gabyx** [comment]
OK this push should work now :-)

**Comment by gabyx** [comment]
Ok, now almost all tests worked, but only the LFS test hanged -> Wiht some investigation I found, its really bad to call `git pull` and `git clone`  (git hooks update shared`) without really resetting the environment -> so the only good and robust solution is `env -i git clone` . Since a `git clone` inside `cli.sh` would run hooks `base-template.sh`-`post-checkout` eventually calling `git lfs` and leaking GIT_DIR variables and probably other variables really affect these commands.  So I guarded now these with `env -i` where happily all test work now. Strange thing...

**Comment by gabyx** [comment]
The original `is_running_in_git_repo_root()` already worked everywhere inside a git repo... we can also
rename it to something else: Suggestion?

File: base-template.sh [hunk 11]
```diff
@@ -703,10 +703,16 @@ show_prompt() {
         # else: Running fallback...
     fi
 
-    # Read from stdin
+    # Read from stdin if its available
     printf "%s %s [%s]:" "$TEXT" "$HINT_TEXT" "$SHORT_OPTIONS"
-    # shellcheck disable=SC2229
-    read -r "$VARIABLE" </dev/tty
+    if [ -f /dev/tty ]; then
```
Comment (rycus86) [inline base-template.sh]: This should be done by the test setup, interactively we expect you have TTY elsewhere. Did you leave this change in by accident during testing?

File: cli.sh [hunk 12]
```diff
@@ -96,20 +96,31 @@ set_main_variables() {
...
 # Returns:
 #   0 if it is likely a Git repository,
 #   1 otherwise
 #####################################################
 is_running_in_git_repo_root() {
-    if ! git status >/dev/null 2>&1; then
-        return 1
-    fi
-
+    CURRENT_GIT_DIR_IS_BARE=$(git rev-parse --is-bare-repository)
```
Comment (rycus86) [inline cli.sh]: This should be just `git rev-parse || return 1` with redirecting output to devnull. We want to answer here whether we're in a git repo or not, it shouldn't care about bare or not, and should also not have side effects of setting other variables.
Not sure what's `[ -d "${CURRENT_GIT_DIR}" ] || return 1` doing right now, maybe that needs to go now?

File: cli.sh [hunk 13]
```diff
@@ -96,20 +96,31 @@ set_main_variables() {
...
 
+#####################################################
+# Checks CURRENT_GIT_DIR_IS_BARE if set to `true`.
+
+# Returns:
+#   0 if it is a bare Git repository,
+#   1 otherwise
+#####################################################
+is_bare_repo() {
+    [ "$CURRENT_GIT_DIR_IS_BARE" = "false" ] || return 1
```
Comment (rycus86) [inline cli.sh]: This should be something like this perhaps?
```
if ! git rev-parse --is-bare-repository 2>/dev/null | grep true; then
    return 1
fi
```

Comment (rycus86) [reply cli.sh]: Or better yet, read it into a variable and do an equals test for it.

File: cli.sh [hunk 14]
```diff
@@ -399,13 +410,18 @@ git hooks trust [forget]
...
             touch .githooks/trust-all &&
-            git config githooks.trust.all Y &&
-            echo "The current repository is now trusted." &&
-            echo "  Do not forget to commit and push the trust marker!" &&
-            return
-
-        echo "! Failed to mark the current repository as trusted" >&2
-        exit 1
+            git config githooks.trust.all Y
+        if ! mkdir -p .githooks ||
```
Comment (rycus86) [inline cli.sh]: Why is this change necessary?
We seem to be repeating what's already been done.

File: cli.sh [hunk 15]
```diff
@@ -856,15 +876,16 @@ add_shared_hook_repo() {
...
 
-        echo "# Added on $(date)" >>"$(pwd)/.githooks/.shared" &&
-            echo "$SHARED_REPO_URL" >>"$(pwd)/.githooks/.shared" &&
-            echo "The new shared hook repository is successfully added" &&
-            echo "  Do not forget to commit the change!" &&
-            return
-
-        echo "! Failed to add the new shared hook repository" >&2
-        exit 1
+        if ! echo "# Added on $(date)" >>"$(pwd)/.githooks/.shared" ||
```
Comment (rycus86) [inline cli.sh]: There seem to be a few changes like this for bare repos.
I'd prefer leaving the previous print chains as-is, and perhaps add a new function: `echo_if_not_bare_repo ${msg}`
That could then suppress printing the message if not needed for bare repos.

On a side note: does printing these messages on a bare repo cause any issues?

File: install.sh [hunk 16]
```diff
@@ -3947,18 +3994,24 @@ install_into_existing_repositories() {
 
     git config --global githooks.previous.searchdir "$RAW_START_DIR"
 
-    LOCAL_REPOSITORY_LIST=$(find "$START_DIR" -type d -name .git 2>/dev/null)
+    REPOSITORY_LIST=$(
```
Comment (rycus86) [inline install.sh]: Why did the variable name change here by the way?

File: install.sh [hunk 17]
```diff
@@ -3947,18 +3994,24 @@ install_into_existing_repositories() {
...
+        REPOSITORY_LIST=$(echo "$REPOSITORY_LIST" | sort)
     fi
 
     IFS="$IFS_NEWLINE"
-    for EXISTING in $LOCAL_REPOSITORY_LIST; do
+    for EXISTING in $REPOSITORY_LIST; do
         unset IFS
 
         EXISTING_REPO_ROOT=$(dirname "$EXISTING")
+        if ! is_git_repo "${EXISTING_REPO_ROOT}"; then
```
Comment (rycus86) [inline install.sh]: This is running on the `dirname` of the repo path, how is that going to work with absolute paths?
Shouldn't this be `if ! is_git_repo "${EXISTING}"; then` ?

File: install.sh [hunk 18]
```diff
@@ -3977,7 +4030,12 @@ install_into_existing_repositories() {
 ############################################################
 install_hooks_into_repo() {
     TARGET_ROOT="$1"
-    TARGET="${TARGET_ROOT}/.git"
+    TARGET="${TARGET_ROOT}"
+    IS_BARE=$(cd "${TARGET_ROOT}" && git rev-parse --is-bare-repository)
+
+    if [ "${IS_BARE}" = "false" ]; then
```
Comment (rycus86) [inline install.sh]: Maybe personal preference, but this would read nicer for me:
```
if [ "${IS_BARE}" = "true" ]; then
    TARGET="${TARGET_ROOT}"
else
    TARGET="${TARGET_ROOT}/.git"
fi
```

File: tests/step-110.sh [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+
+echo "Test hook from push"
+cd /tmp/test110/local || exit 1
+echo "Test" >Test
+git add Test || exit 1
+git commit -a -m "First" || exit 1
+echo "Push hook to fail"
+OUTPUT=$(git push 2>&1)
+
+#shellcheck disable=1014,2181
```
Comment (rycus86) [inline tests/step-110.sh]: Is it possible to resolve these issues rather than disable their checks?

File: uninstall.sh [hunk 20]
```diff
@@ -137,13 +143,21 @@ uninstall_from_existing_repositories() {
 #   1 otherwise
 #####################################################
 is_running_in_git_repo_root() {
-    if ! git status >/dev/null 2>&1; then
-        return 1
-    fi
-
+    git rev-parse >/dev/null 2>&1 || return 1
```
Comment (rycus86) [inline uninstall.sh]: Shouldn't this be `is_git_repo || return 1` ?

File: uninstall.sh [hunk 21]
```diff
@@ -157,7 +171,7 @@ uninstall_from_current_repository() {
         exit 1
     fi
 
-    uninstall_hooks_from_repo "$CURRENT_GIT_DIR"
+    uninstall_hooks_from_repo "$(pwd)"
```
Comment (rycus86) [inline uninstall.sh]: Why this change?

File: tests/step-110.sh [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+    exit 1
+fi
+
+mkdir -p /tmp/test110/hooks &&
+    mkdir -p /tmp/test110/server &&
+    mkdir -p /tmp/test110/local || exit 1
+
+# Hooks
+cd /tmp/test110/hooks && git init || exit 1
+/var/lib/githooks/uninstall.sh --local || exit 1
```
Comment (rycus86) [inline tests/step-110.sh]: Why do we need to uninstall Githooks here?

**Comment by rycus86** [comment]
Thanks for the changes so far! I left some more comments.

I still have concerns about this PR in general: it feels like this might be a pretty niche use-case that most people won't need, or you could only apply it to a very small number of repos where it might make sense - would you agree?

I'm thinking that maybe you'd be better off with a simple hook on bare repos without Githooks in the mix - should we explore how that might look like?

I'm trying to establish 1) what value this PR and Githooks would give you on bare repos, 2) how commonly do you need this on your bare repos, 3) is it worth the complexity?
Thanks!

**Comment by gabyx** [comment]
With my changes now, which are not really much after your review I think the value is the following:
Since we would like to use Githooks company-wide, where all its update behavior come in handy, since we can manage hooks at a central place. It would be also very handy to use the exact same behavior on our repositories on the server, of course the update mechanism might be disabled and update is needed to be done manually by `git hooks update`or `git hooks shared update` since people are working in parallel (`git push`) and I dont know how that works out...
But the thing is, git has bare  and non-bare repositories, and why should Githooks not support this. Every hook manager really should, most don't, and thats cumbersome, because there is not really much of a difference considering hooks in a bare and non-bare repo. 
I wouldn't think there is now much more complexity for Githooks. The changes with `env -i` and stuff are actually strange things which really can happen also without the support for bare repos, they just came up in the test procedure, and made me puzzle and worrying about why the hack -> thinking more deeply about it goes into the direction that "are most git invocations really safe here, or might it do stupid stuff". However, we can always of course copy our own hooks to the repository of course without having githooks installed on the server ;-). The good thing is. When you make new repos on lets say Gitea, in the background the githook manager gets installed directyl from the templates and everything is setup, meaning global shared hooks are run for example and so forth :-). Its really handy I think :-). Thats why I'd think it would be valuable to have this. otherwise, we are back to managing global shared hooks manually on gitea when adding a new repo, Meaning -> People need to contact the admin -> the admin logs into the container, makes the adjustments, with githooks this is not needed anymore :-)

**Comment by gabyx** [comment]
Or thats basically my motivation behind it :-). If it works out, I will see, so far, we are testing the hooks manager with a few people before we distribute a global company wide install script, which sets up everything nicely for everybody including githooks :-)

**Comment by rycus86** [comment]
OK, thanks for the explanation! You're right, the latest changes don't seem much and should improve things anyway, so let's do it!

I have a few minor things left plus some unresolved comments, I'll mark those up for you to fix then it should be good to go. 🎉

Please also make sure your final commit message reflects this PR rather than "Fixes". :)

Thanks a lot for your work on this!

File: README.md [hunk 23]
```diff
@@ -5,7 +5,7 @@
 
 A simple Shell script to support per-repository [Git hooks](https://git-scm.com/docs/githooks), checked into the actual repository that uses them.
 
-To make this work, it creates hook templates that are installed into the `.git/hooks` folders automatically on `git init` and `git clone`. When one of them executes, it will try to find matching files in the `.githooks` directory under the project root, and invoke them one-by-one. There's more to the story though, you can read about it under the [Templates or global hooks](#Templates-or-global-hooks) section.
+To make this work, it creates hook templates that are installed into the `.git/hooks` folders automatically on `git init` and `git clone` or into `./hooks` for bare repositories. When one of them executes, it will try to find matching files in the `.githooks` directory under the project root, and invoke them one-by-one. There's more to the story though, you can read about it under the [Templates or global hooks](#Templates-or-global-hooks) section.
```
Comment (rycus86) [inline README.md]: Please revert this, here in the intro it's just confusing I think, it's enough to mention bare repos later.

File: README.md [hunk 24]
```diff
@@ -118,7 +118,7 @@ You can also manage and update shared hook repositories using the [command line
 
 ## Opt-in hooks
 
-To try and make things a little bit more secure, Githooks checks if any new hooks were added we haven't run before, or if any of the existing ones have changed. When they have, it will prompt for confirmation whether you accept those changes or not, and you can also disable specific hooks to skip running them until you decide otherwise. The accepted checksums are maintained in the `.git/.githooks.checksum` file, per local repository.
+To try and make things a little bit more secure, Githooks checks if any new hooks were added we haven't run before, or if any of the existing ones have changed. When they have, it will prompt for confirmation whether you accept those changes or not, and you can also disable specific hooks to skip running them until you decide otherwise. The accepted checksums are maintained in the `.git/.githooks.checksum` file per local repository (or `./.githooks.checksum` file per bare repositories).
```
Comment (rycus86) [inline README.md]: Please revert this.

File: README.md [hunk 25]
```diff
@@ -219,6 +219,19 @@ $ sh -c "$(curl -fsSL https://raw.githubusercontent.com/rycus86/githooks/master/
 
 The GitHub URL also accepts the additional parameters mentioned above, the `https://r.viktoradam.net/githooks` URL is just a redirect to the longer GitHub address.
 
+#### Setup for bare repositories
```
Comment (rycus86) [inline README.md]: This one is good

File: README.md [hunk 26]
```diff
@@ -233,7 +246,7 @@ When one of them executes, it will try to find matching files in the `.githooks`
 
 #### Template folder
 
-In this approach, the install script creates hook templates that are installed into the `.git/hooks` folders automatically on `git init` and `git clone`.
+In this approach, the install script creates hook templates that are installed into the `.git/hooks` folders automatically on `git init` and `git clone`. For bare repositories, the hooks are installed into the `./hooks` folder on `git init --bare`.
```
Comment (rycus86) [inline README.md]: This is also good.

File: base-template.sh (hunk 11 above)
Comment (rycus86) [reply base-template.sh]: Please check this one, I'm not sure this change was needed.

File: cli.sh [hunk 27]
```diff
@@ -97,19 +97,28 @@ set_main_variables() {
...
 #   0 if it is likely a Git repository,
 #   1 otherwise
 #####################################################
 is_running_in_git_repo_root() {
-    if ! git status >/dev/null 2>&1; then
-        return 1
-    fi
-
+    git rev-parse >/dev/null 2>&1 || return 1
     [ -d "${CURRENT_GIT_DIR}" ] || return 1
```
Comment (rycus86) [inline cli.sh]: Please double-check that this still makes sense for bare repos.
The `${CURRENT_GIT_DIR}` variable is set to `git rev-parse --common-dir` or similar, no idea if that returns  something sensible for bare repos. /final

File: install.sh (hunk 17 above)
Comment (rycus86) [reply install.sh]: Please check this I think this is a bug. /final

File: install.sh [hunk 28]
```diff
@@ -3977,7 +4011,13 @@ install_into_existing_repositories() {
 ############################################################
 install_hooks_into_repo() {
     TARGET_ROOT="$1"
-    TARGET="${TARGET_ROOT}/.git"
+    
+    IS_BARE=$(cd "${TARGET_ROOT}" && git rev-parse --is-bare-repository)
```
Comment (rycus86) [inline install.sh]: Could you send errors to devnull? `2>/dev/null` /final

File: tests/step-110.sh (hunk 19 above)
Comment (rycus86) [reply tests/step-110.sh]: What about this one? /final

File: uninstall.sh [hunk 29]
```diff
@@ -121,9 +121,15 @@ uninstall_from_existing_repositories() {
...
     fi
 
-    find "$START_DIR" -type d -name .git 2>/dev/null | while IFS= read -r EXISTING; do
-        uninstall_hooks_from_repo "$EXISTING"
-    done
+    find "$START_DIR" \( -type d -and -name .git \) -or \
+        \( -type f -and -name HEAD -and -not -path "*/.git/*" \) 2>/dev/null |
+        while IFS= read -r EXISTING; do
+            EXISTING_REPO_ROOT=$(dirname "$EXISTING")
+            if ! is_git_repo "${EXISTING_REPO_ROOT}"; then
```
Comment (rycus86) [inline uninstall.sh]: Same as in the install script, this would pass in the directory name, not the path of the directory. /final

File: uninstall.sh [hunk 30]
```diff
@@ -168,7 +182,15 @@ uninstall_from_current_repository() {
 #   None
 ############################################################
 uninstall_hooks_from_repo() {
-    TARGET="$1"
+    TARGET_ROOT="$1"
+
+    IS_BARE=$(cd "${TARGET_ROOT}" && git rev-parse --is-bare-repository)
```
Comment (rycus86) [inline uninstall.sh]: Please send errors to devnull here too. /final

**Comment by rycus86** [comment]
Added a `/final` suffix to my remaining comments to help you search for them here.

**Comment by gabyx** [comment]
Hi; =)
- Changed all with 👍 
- `EXISTING_REPO_ROOT=$(dirname "$EXISTING")` # EXISTING = "/root/bla/repo/.git"
`dirname`  returns the directory path (parent path)... so its correct.  = `/root/bla/repo`. Thats what the commands should receive...
- `if [ -f /dev/tty ]; then` is still needed to suppress error output : "file descriptor not found bla bla". Of course it works, but is just strange output.

**Comment by gabyx** [comment]
Possible to merge? Thx :-)

**Comment by rycus86** [comment]
Sorry, haven't had a chance to look at this yet, I'll try this week. Thanks for your patience and all the changes! :)

**Comment by rycus86** [comment]
Sorry again for the delay and thanks for your patience!
One of these days I will manage to get to my laptop and review these PRs, not only on my phone...

**Comment by rycus86** [comment]
Thanks! I was just looking at this, I think it's good to go.
There are a few `/final` comments you haven't replied yet, can you just double-check them amd reply before I merge?
Ignore the one about `dirname`, you explained and I realized I thought wrong what that does. :)

**Comment by rycus86** [comment]
Let me know when you finished with this, then I'll get it in master. :tada:

**Comment by gabyx** [comment]
Thx. I replied. Maybe check the `install_` and `uninstall_` function again. Maybe you see some errors I made...

**Comment by gabyx** [comment]
FYI: This is ready.

**Comment by rycus86** [comment]
Thanks a lot! I'll try to review it in an hour or so.

File: base-template.sh (hunk 11 above)
Comment (rycus86) [reply base-template.sh]: Out of curiosity: did you need this change for sure?
If not, I might revert, I think other places would have the same read with tty redirection.

File: install.sh [hunk 31]
```diff
@@ -4028,44 +4067,53 @@ install_hooks_into_repo() {
     done
 
     # Offer to setup the intro README if running in interactive mode
-    if is_non_interactive; then
-        true # Let's skip this in non-interactive mode to avoid polluting the local repos with README files
+    # Let's skip this in non-interactive mode or in a bare repository
+    # to avoid polluting the repos with README files
+    if ! is_non_interactive && [ "${IS_BARE}" != "true" ]; then
```
Comment (rycus86) [inline install.sh]: Nit: this chain is getting a bit big, it would be good to change to something like
```
if is_non_interactive; then
  true
elif [ "${IS_BARE}" != "true" ]; then
  true
else
  ...
fi
```
I'm happy to do this one after merging if you don't have time, or not at a computer.

File: install.sh [hunk 32]
```diff
@@ -4028,44 +4067,53 @@ install_hooks_into_repo() {
...
+
+        # Getting the working tree (no external .git directories)
+        # see https://stackoverflow.com/a/38852055/293195
```
Comment (rycus86) [inline install.sh]: Out of curiosity, why did you need this? What was wrong with the old code?

**Comment by gabyx** [comment]
Made the changes, reverted the /dev/tty.

**Comment by rycus86** [comment]
Thanks a lot for your hard work on this!
